client/client_test.go

```diff
@@ -617,25 +619,32 @@ func TestClient_SaveRestoreState(t *testing.T) {
 	c1.config.PluginSingletonLoader = singleton.NewSingletonLoader(logger, c1.config.PluginLoader)
 
 	c2, err := NewClient(c1.config, consulCatalog, nil, mockService, nil)
-	if err != nil {
-		t.Fatalf("err: %v", err)
-	}
+	must.NoError(t, err)
 	defer c2.Shutdown()
 
 	// Ensure the allocation is running
-	testutil.WaitForResult(func() (bool, error) {
-		c2.allocLock.RLock()
-		ar := c2.allocs[alloc1.ID]
-		c2.allocLock.RUnlock()
-		status := ar.Alloc().ClientStatus
-		alive := status == structs.AllocClientStatusRunning || status == structs.AllocClientStatusPending
-		if !alive {
-			return false, fmt.Errorf("incorrect client status: %#v", ar.Alloc())
-		}
-		return true, nil
-	}, func(err error) {
-		t.Fatalf("err: %v", err)
-	})
+	must.Wait(t, wait.InitialSuccess(
+		wait.ErrorFunc(func() error {
+			c2.allocLock.RLock()
+			defer c2.allocLock.RUnlock()
+
+			_, corruptStillLives := c2.allocs[corruptAlloc.ID]
+			if corruptStillLives {
+				return fmt.Errorf("corrupted allocation should not have been restored")
+			}
```

Comment:
Note to reviewers: I've updated this test to the current idioms, but this is the meat of the change to ensure we handle the corrupted alloc state correctly.